controller/setting_controller.go

```diff
@@ -1501,6 +1504,9 @@ func (info *ClusterInfo) collectVolumesInfo() error {
 		replicaZoneSoftAntiAffinity := info.collectSettingInVolume(string(volume.Spec.ReplicaZoneSoftAntiAffinity), string(longhorn.ReplicaZoneSoftAntiAffinityDefault), types.SettingNameReplicaZoneSoftAntiAffinity)
 		replicaZoneSoftAntiAffinityCountStruct[util.StructName(fmt.Sprintf(ClusterInfoVolumeReplicaZoneSoftAntiAffinityCountFmt, util.ConvertToCamel(string(replicaZoneSoftAntiAffinity), "-")))]++
 
+		replicaDiskSoftAntiAffinity := info.collectSettingInVolume(string(volume.Spec.ReplicaDiskSoftAntiAffinity), string(longhorn.ReplicaDiskSoftAntiAffinityDefault), types.SettingNameReplicaDiskSoftAntiAffinity)
+		replicaDiskSoftAntiAffinityCountStruct[util.StructName(fmt.Sprintf(ClusterInfoVolumeReplicaDiskSoftAntiAffinityCountFmt, util.ConvertToCamel(string(replicaDiskSoftAntiAffinity), "-")))]++
+
```

Comment:
Looks like we have a bug here @c3y1huang 

This setting `replicaDiskSoftAntiAffinity` only has 3 possible value: `ignore`, `enable`, `disable`. However, the upgrade responder might received 5 possible value returns by the function `collectSettingInVolume` which are: `ignore`, `enable`, `disable`, `true`, and `false`

Reply:
This pattern shows up in other storage class parameters.  The global setting is either true or false; the storage class flavor is one of "ignore" (use the global setting), "enable" (force to true), or "disable" (force to false).
I have been supposing that the override-or-not check is made when a storage class is evaluated for use in a volume creation, and any such parameter would be resolved to either true or false in the volume itself, and those would be the only legal values for a volume.

Reply:
Phan explained to me that the parameter like "ignore" is not evaluated at volume creation.  It is retained in the volume and evaluated when needed, which is to say any time a new replica needs to be placed.  At that point, the controller will use whatever the current global setting is.  So it is possible for a volume that doesn't express a preference to change its behavior if the global setting is changed between attempts to place another replica.

Reply:
I did not understand this comment until I started trying to test my changes with the upgrade responder.

The standard configuration for the upgrade responder is to store true or false counts.

```
> show field keys
...
longhorn_volume_replica_auto_balance_disabled_count                 float
longhorn_volume_replica_disk_soft_anti_affinity_true_count          float
longhorn_volume_replica_soft_anti_affinity_false_count              float
```

When I was experimenting, I got the following rows:

```
> select app_version, value, longhorn_setting_replica_disk_soft_anti_affinity, longhorn_volume_replica_disk_soft_anti_affinity_true_count from upgrade_request
name: upgrade_request
time                app_version value longhorn_setting_replica_disk_soft_anti_affinity longhorn_volume_replica_disk_soft_anti_affinity_true_count
----                ----------- ----- ------------------------------------------------ ----------------------------------------------------------
1692199982764244311 v1.6.0-dev  1                                                      
1692199984374879541 v1.6.0-dev  1     true                                             
1692199984932156839 v1.6.0-dev  1                                                      
1692201702247618751 v0.8.1      1                                                      
1692207182563500432 v1.6.0-dev  1                                                      
1692207184123366093 v1.6.0-dev  1     true                                             2
1692207184676124001 v1.6.0-dev  1                                                      
1692210782632842772 v1.6.0-dev  1                                                      
1692210784275448343 v1.6.0-dev  1     false                                            
1692210784788222835 v1.6.0-dev  1                                                      
1692211523658344337 v1.6.0-dev  1                                                      
1692211528320057508 v1.6.0-dev  1                                                      
1692211534450305082 v1.6.0-dev  1     false        
```

I got the row with the 2 in it when the `ReplicaDiskSoftAntiAffinity` setting was `true` and both volumes had `ignored`.

The [collectSettingInVolume function](https://github.com/longhorn/longhorn-manager/blob/c144fc16d8b30f1eda63005bae83e6604586554a/controller/setting_controller.go#L1560-L1569) counted both volumes as `true` (which comes from the `ReplicaDiskSoftAntiAffinity` setting itself.

However, when I set the `ReplicaDiskSoftAntiAffinity` setting to `false` and one volume to `enabled` before the last line, presumably one volume was counted as `enabled` and one as `false`. Since neither are `true` they are not represented in the database.

@PhanLe1010, I think there are only four possible values: `true`, `false`, `enabled`, and `disabled` right now. If a volume has ignored, it will register as `true` or `false`.

Not sure exactly what we want to see here. IMO, we actually want to reduce the data to a boolean, whether we represent it as `enabled` vs. `disabled` or `true` vs. `false`. We want to know whether or not the volume is actually using `ReplicaDiskAntiAffinity` at the moment, not necessarily the exact value of this field setting.

Reply:
> Not sure exactly what we want to see here. IMO, we actually want to reduce the data to a boolean, whether we represent it as enabled vs. disabled or true vs. false. We want to know whether or not the volume is actually using ReplicaDiskAntiAffinity at the moment, not necessarily the exact value of this field setting.

Agree with this one. Could we create a ticket for this since it is happening with other settings also?

Reply:
https://github.com/longhorn/longhorn/issues/6593